paddle/framework/recurrent_network_op.h

```diff
@@ -0,0 +1,141 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#pragma once
+
+#include "paddle/framework/enforce.h"
+#include "paddle/framework/scope.h"
+#include "paddle/framework/variable.h"
+
+namespace paddle {
+namespace framework {
+
+// fake interfaces that has not be implemented by other modules.
+struct OpRunContext {
+  Scope* scope;
+};
+
+// TODO replace this with Net's proto.
+struct NetDesc {
+  std::string name;
+}
+
+class OperatorBase {
+ public:
+  virtual ~OperatorBase() {}
```

Comment:
I looks to me that the constructor needs a parameter `paddle::framework::proto::OperatorDesc` so could it possble to call `InferShape`, which saves sizes of inputs/outputs into the desc.  Only if so, we could have all necessary information for calling `OperatorBase::Run`:

```cpp
class OperatorBase {
 public:
  OperatorBase(const proto::OperatorDesc& desc) : desc_(desc) {}
  virtual void Run(OpRunContext* context) const = 0;

 protected:
  virtual void InferShape(const Scope* scope) const = 0; // needs to read from and write to desc_

  proto::OperatorDesc desc_;
};
```

So the information in `proto::OperatorDesc` propagates along the path:

```
Operator's constructor 
  ① ↓
OperatorBase::desc_  → Operator's Run
  ②↓ ↑③            ④
Operator's InferShape
```

@Superjom @reyoung @jacquesqiao